### DISCUSSION TITLE
estimating the quality of PCE expansion

### DISCUSSION THREAD

**Discussion by cscpdb30** [snl-dakota/dakota#80] [Q&A]
I'm using Dakota to extract Sobol indices via a Polynomial Chaos Expansion. I would like to check the quality of the expansion.
The output includes the PCE coefficients but using them entails the building of the polynomials and additional work when Dakota could do that directly.

So my question: is there a way to make Dakota provide a measure of the quality of the expansion? As I understand, there are ways to do this when using _p_refinement_ or the keyword _metrics_ when using surrogate models.
Can I use _metrics_ with a _polynomial_chaos_ method with a fixed _expansion_order_?

**Comment by Trogdor-Forever** [comment]
Following. I am also using PCE to obtain Sobol indices from coefficient approximations using point collocation sparse regression methods that use a fixed expansion order. Look under "method-->polynomial_chaos-->expansion_order-->collocation_points-->regression algorithm type" (note: Sandia needs to update their online documentation under the expansion_order link to include these extra options https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/reference/method-polynomial_chaos-expansion_order.html, but you can find them in the using the nested link tree on the left side of the webpage). 

What I do is incrementally batch sample evaluations with Dakota and then trick Dakota into regressing subset batch evaluations by using a dummy driver to feed the 'fork' params and results files for a given batch size. That way, you can see how the Sobol indices and even moments change as sample sizes increase for a given regression algorithm. If you use MATLAB, I'd highly recommend UQLab's software as this has a much more intuitive (yet more limited) interface that includes leave one out (LOO) errors and normalized empirical errors as evaluation batch sizes increase. I also feed Dakota's inputs/outputs into UQLab to compare their results and Sobol index / other error convergence rates. 

Aside: So far, I've found that for 200 evaluations, PCE order of 2, and 195 parameters, most of Dakota's sparse regression algorithms drive down the number interaction Sobol indices so the sum of all the parameters' total Sobol indices is essentially 1.0 when compared to UQLab's, so maybe I'm reading the wrong output values.
![image](https://github.com/snl-dakota/dakota/assets/120498230/7eaddf3a-909c-414e-bd03-e1ca19a0c98e)

**Reply by cscpdb30** [reply]
thank you very much. what model are you using to do that? I mean, it is not single, is it?

**Comment by Trogdor-Forever** [comment]
> thank you very much. what model are you using to do that? I mean, it is not single, is it?
No I do not use model, below is what I use for input. I loop through this input file with another bash script to change the chosen sparse regression algorithm and number of collocation points. Note the "seed" tag and "allow_existing_results" tag under fork. These appear to allow Dakota to use the same parameter inputs and outputs that were generated in a previous run. There are 200 params.in.* and 200 results.out.* files. All my driver does is pass the parameters generated from the same "seed" and links it with a results file that already exists in the working directory to trick Dakota into doing the method(s). There's probably a better way to do this with "import_build_points_file" under method, but it was giving me garbage responses when I tried. This way works for me.

Input file:
"method,
    polynomial_chaos 
        sample_type lhs
        seed = 15347
        expansion_order = 2 
			collocation_points = 200
				basis_pursuit_denoising
                                  #orthogonal_matching_pursuit
                                  #basis_pursuit
                                  #basis_pursuit_denoising
                                  #least_angle_regression
                                  #least_absolute_shrinkage
                                  #cross_validation
                                            #noise_only
        variance_based_decomp


#model
#single

variables
    loguniform_uncertain = 195
        upper_bounds 2.10E+22 ... 1.17E+34
        initial_point 7.00E+21 ... 3.90E+33
        lower_bounds 2.10E+21 ... 1.56E+33
        descriptors 'CHEM0001' ... 'CHEM0195'

interface
    analysis_drivers = 'driver.sh'
        fork
            parameters_file = 'params.in'
            results_file = 'results.out'
            allow_existing_results
            file_tag
            file_save

responses
  descriptors 'IRA' 'IRB' 'IRC' 'IRT' 
  response_functions = 4
    scalar_responses = 0
    field_responses = 4
        lengths = 175 175 175 175
  no_gradients
  no_hessians"

Driver file:
"#!/bin/bash

params=$1
results_file=$2

num=`echo $params | cut -c 11-`

mv results.out.$num results.out.temp
cp results.out.temp $results_file"
